apps/ledger-live-mobile/src/screens/FirmwareUpdate/useUpdateFirmwareAndRestoreSettings.ts

```diff
@@ -0,0 +1,124 @@
+  device: Device;
+  updateFirmwareAction?: (
+    args: updateFirmwareActionArgs,
+  ) => Observable<UpdateFirmwareActionState>;
+};
+
+export type UpdateStep =
+  | "start"
+  | "imageBackup"
+  | "firmwareUpdate"
+  | "languageRestore"
+  | "imageRestore"
+  | "appsRestore"
+  | "completed";
+
+const staxLoadImageAction = createStaxLoadImageAction(staxLoadImage);
+const staxFetchImageAction = createStaxFetchImageAction(staxFetchImage);
+
+export const useUpdateFirmwareAndRestoreSettings = ({
+  updateFirmwareAction,
+  device,
+}: FirmwareUpdateParams) => {
+  const [updateStep, setUpdateStep] = useState<UpdateStep>("start");
+
+  const staxFetchImageRequest = useMemo(() => ({}), []);
+  const staxFetchImageState = staxFetchImageAction.useHook(
+    updateStep === "imageBackup" ? device : null,
+    staxFetchImageRequest,
+  );
+
+  const { triggerUpdate, updateState: updateActionState } = useUpdateFirmware({
+    deviceId: device?.deviceId ?? "",
+    updateFirmwareAction,
+  });
+
+  const staxLoadImageRequest = useMemo(
+    () => ({
+      hexImage: staxFetchImageState.hexImage ?? "",
+      padImage: false,
+    }),
+    [staxFetchImageState.hexImage],
+  );
+  const staxLoadImageState = staxLoadImageAction.useHook(
+    updateStep === "imageRestore" && staxFetchImageState.hexImage
+      ? device
+      : null,
+    staxLoadImageRequest,
+  );
+
+  useEffect(() => {
+    switch (updateStep) {
+      case "start":
+        if (device.modelId === DeviceModelId.stax) {
+          setUpdateStep("imageBackup");
+        } else {
+          setUpdateStep("firmwareUpdate");
+        }
+        break;
+      case "imageBackup":
+        if (staxFetchImageState.imageFetched || staxFetchImageState.error) {
+          // TODO: check if we want to do something with error, maybe just log it
+          setUpdateStep("firmwareUpdate");
+        }
+        break;
+      case "firmwareUpdate":
+        if (updateActionState.step === "preparingUpdate") {
+          triggerUpdate();
+        } else if (updateActionState.step === "firmwareUpdateCompleted") {
+          setUpdateStep("languageRestore");
+        }
+        break;
+      case "languageRestore":
+        // TODO: implement image restore
+        setUpdateStep("imageRestore");
+        break;
+      case "imageRestore":
+        if (
+          staxLoadImageState.imageLoaded ||
+          staxLoadImageState.error ||
+          !staxFetchImageState.hexImage
+        ) {
+          // TODO: check if we want to do something with error, maybe just log it
+          setUpdateStep("appsRestore");
+        }
+        break;
+      case "appsRestore":
+        setUpdateStep("completed");
+        break;
+      default:
+        break;
+    }
+  }, [
+    device.modelId,
+    staxFetchImageState.error,
+    staxFetchImageState.imageFetched,
+    staxLoadImageState.error,
+    staxLoadImageState.imageLoaded,
+    triggerUpdate,
+    updateActionState.step,
+    updateStep,
+  ]);
```

Comment:
it's missing the `staxFetchImageState.hexImage` in the dep ;)

Reply:
Good catch, fixed it.